docs: dev guide updates (#711)
bbcmicrobit/micropython: docs/devguide/flashfirmware.rst

.. _flashfirmware:

========================

> Build and flash firmware

Reviewer: Let's rename this (and the link on the left menu) to say `Build and flash firmware`.
We've used the word "firmware" very early on the MicroPython nomenclature, but it can be used with the use of "firmware" in the rest of the micro:bit documentation referring to the DAPLink firmware.

Reviewer: @microbit-mark sorry, I meant `Build and flash MicroPython` here, as I wanted to use the word firmware too much.

Author: Done in https://github.com/bbcmicrobit/micropython/pull/711/commits/939efde1f837cd9aea60e972e3bb135b215baff6

========================

micro:bit V1
============

This applies to MicroPython for the micro:bit V1, the source of which can be 
found here: `bbcmicrobit/micropython <https://github.com/bbcmicrobit/micropython>`_.

Dependencies
------------
- `cMake <https://cmake.org/>`_

> - `gcc <http://gcc.gnu.org/install/>`_

Reviewer: Done.

- `git <https://git-scm.com/>`_
- `ninja <https://ninja-build.org/>`_
- `python <https://www.python.org/downloads/>`_

> - `yotta <http://docs.yottabuild.org//>`_

Reviewer: Done,

The `yotta
<http://docs.yottabuild.org//>`_ tool is used to build MicroPython, but before
that takes place additional files have to be generated by the Makefile in
preparation for the build, and additional data is added to the hex file after.

> Use yotta to build::

Reviewer: Done.

>   yt target bbc-microbit-classic-gcc@https://github.com/lancaster-university/yotta-target-bbc-microbit-classic-gcc

Reviewer: This doesn't look right, I think it should be the `-nosd` target?
Did this come from other MicroPython instructions? In that case we'll need to update those as well?

Author: I think it was my assumption based on https://support.microbit.org/support/solutions/articles/19000123284-yotta-registry-deprecation I will update the target

Reviewer: Done.

.. _flashfirmware:

========================
Build and flash firmware
========================

micro:bit V1
============

This applies to MicroPython for the micro:bit V1, the source of which can be 
found here: `bbcmicrobit/micropython <https://github.com/bbcmicrobit/micropython>`_.

Dependencies
------------
- `cMake <https://cmake.org/>`_
- `Arm gcc <https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads>`_
- `git <https://git-scm.com/>`_
- `ninja <https://ninja-build.org/>`_
- `python <https://www.python.org/downloads/>`_
- `srecord <http://srecord.sourceforge.net/>`_
- `yotta <http://docs.yottabuild.org//>`_

The `yotta
<http://docs.yottabuild.org//>`_ tool is used to build MicroPython, but before
that takes place additional files have to be generated by the Makefile in
preparation for the build, and additional data is added to the hex file after.

Clone the repository and change directory to it::

>   $ git clone <https://github.com/bbcmicrobit/micropython
> 

Reviewer: ```suggestion
  $ git clone https://github.com/bbcmicrobit/micropython
```

Author: done as part of https://github.com/bbcmicrobit/micropython/pull/711/commits/939efde1f837cd9aea60e972e3bb135b215baff6

>   $ cd bbcmicrobit/micropython

Reviewer: ```suggestion
  $ cd micropython
```

Author: done as part of https://github.com/bbcmicrobit/micropython/pull/711/commits/939efde1f837cd9aea60e972e3bb135b215baff6

Using ``tools/hexlify.py`` you can turn a Python script into Intel HEX format
to be concatenated at the end of the MicroPython firmware.hex.  A simple header
is added to the script.::

  ./hexlifyscript.py <script.py>
  
It also accepts data on standard input.

micro:bit V2
============

This applies to MicroPython for the micro:bit V2, the source of which can be 
found here: `microbit-foundation/micropython-microbit-v2 <https://github.com/microbit-foundation/micropython-microbit-v2>`_.

The repository also contains a history of 
`MicroPython firmware builds <https://github.com/microbit-foundation/micropython-microbit-v2/actions>`_.

Dependencies
------------

- `Arm gcc <https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads>`_
- `gcc <http://gcc.gnu.org/install/>`_
- `cMake <https://cmake.org/>`_
- `git <https://git-scm.com/>`_ 
- `ninja <https://ninja-build.org/>`_
- `python <https://www.python.org/downloads/>`_

Clone the repository and change directory to it::

    $ git clone https://github.com/microbit-foundation/micropython-microbit-v2
    $ cd micropython-microbit-v2

Update the submodules::

>     $ git submodule update --init

Reviewer: This didn't render correctly:
![image](https://user-images.githubusercontent.com/29712657/110932630-a3322a80-8323-11eb-9145-a6599483f125.png)

Please remember you can check the RtD PR build after each push.

Author: fixed in https://github.com/bbcmicrobit/micropython/pull/711/commits/939efde1f837cd9aea60e972e3bb135b215baff6

    $ git submodule update --init

Then build the MicroPython cross-compiler::

    $ make -C lib/micropython/mpy-cross

After setting up, go to the src/ directory and build::

    $ cd src

    $ make

The resulting firmware will be ``MICROBIT.hex`` in the ``src/`` 

> directory whichcan be copied to the micro:bit.

Reviewer: ```suggestion
directory which can be copied to the micro:bit.
```

.. _flashfirmware:

========================
Build and flash firmware
========================

micro:bit V1
============

This applies to MicroPython for the micro:bit V1, the source of which can be 
found here: `bbcmicrobit/micropython <https://github.com/bbcmicrobit/micropython>`_.

> Dependencies

Reviewer: I think these instructions depend on the Yotta installer, which might disappear and even when it was maintained it was usually outdated, we should update these to list the tools the users will need to install.

Reviewer: done

------------
- `yotta <http://docs.yottabuild.org//>`_

When installing `yotta <http://docs.yottabuild.org//>`_, make sure you have
these components ticked to install.

- python
- gcc
- cMake
- ninja
- Yotta
- git-scm

> - mbed serial driver

Reviewer: Done

The `yotta
<http://docs.yottabuild.org//>`_ tool is used to build MicroPython, but before that takes place
additional files have to be generated by the Makefile in preparation for the 
build, and additional data is added to the hex file after.

Use yotta to build::

  yt target bbc-microbit-classic-gcc@https://github.com/lancaster-university/yotta-target-bbc-microbit-classic-gcc

Run yotta update to fetch remote assets::

  yt up

The yotta tool is used to build MicroPython, but before that takes place
additional files have to be generated by the Makefile in preparation for the 
build, and additional data is added to the hex file after.

Running the ``make all`` command executes the following steps:

- The ``tools/makeversionhdr.py`` script creates the ``microbitversion.h`` file 
  with macros containing build information
- Yotta builds the source and creates a bare hex file with just the firmware
- The ``tools/adduicr.py`` script adds the UICR to the bare hex
- The final hex file is placed in ``build/firmware.hex``
- The user can optionally append a script using ``tools/makecombinedhex.py`` 

>   (or other tools)

Reviewer: This is very useful info for the Firmware Hex file section, but probably not necessary for most users that just want to build it.
I think the previous content here fit better.

Reviewer: Done.

micro:bit V2

> ============

Reviewer: The dependencies are a bit different for V2, so we should  list them here as well

Reviewer: Done.

  make all
Running the ``make all`` command executes the following steps:

The result is a microbit-micropython hex file (i.e. ``microbit-micropython.hex``)
found in the build/bbc-microbit-classic-gcc-nosd/source from the root of the
repository.
- The ``tools/makeversionhdr.py`` script creates the ``microbitversion.h`` file 
  with macros containing build information
- Yotta builds the source and creates a bare hex file with just the firmware
- The ``tools/adduicr.py`` script adds the UICR to the bare hex
- The final hex file is placed in ``build/firmware.hex``
- The user can optionally append a script using ``tools/makecombinedhex.py`` 
  (or other tools)

The Makefile does some extra preprocessing of the source, which is needed only
if you add new interned strings to ``qstrdefsport.h``. The Makefile also puts
the resulting firmware at build/firmware.hex, and includes some convenience
targets.
micro:bit V2
============

> Preparing firmware and a Python program

Reviewer: This section was still useful for MicroPython v1, we can just indicate the commands needed to ru makecombined

Reviewer: Done.

> underlying target platform.

Reviewer: I think this might raise more questions than it resolves. We should probably either explain what DAL and CODAL are, or remove this bit.

Reviewer: Done.

>     $ git submodule update --init

Reviewer: Let's also add the git clone step to reinforce that this is a different repo.

Reviewer: Done.

That will build both ``libmicropython.a`` (from source in ``src/codal_port/``) and the 
CODAL app (from source in ``src/codal_app/``). The resulting firmware will be 

> ``MICROBIT.hex`` in the ``src/`` directory which can be copied to the micro:bit.

Reviewer: We can probably just mention the hex and lib out the .a files.

Reviewer: @microbit-mark this is still outstanding, but if you think we should leave it in let me know.

Author: I've just left the hex info in https://github.com/bbcmicrobit/micropython/commit/939efde1f837cd9aea60e972e3bb135b215baff6

>   yt build

Reviewer: Done.

...or use the makefile::

  make all

The result is a microbit-micropython hex file
(i.e. ``microbit-micropython.hex``) found in the

> build/bbc-microbit-classic-gcc-nosd/source from the root of the repository.

Reviewer: I think the readme lists `build/firmware.hex`, could you check what's the right path for the final hex?

Author: I can't get the build working on my machine at the moment I get this error when trying to set the target
```
dyld: Library not loaded: @executable_path/../.Python
  Referenced from: /Users/mark/yotta-venv/bin/python3.7
  Reason: image not found
[1]    2649 abort      yt target 
```
I will update to `firmware.hex` in line with the readme.

Reviewer: Sounds like the yotta virtual environment might have been created from a Python installation that no longer exists (or has changed)?
I assume you would get a similar issue when trying to execute any yotta command?
If so, you'll have to recreate the virtual environment with a new one.

Author: Thanks, Carlos. Recreating the virtual environment worked and output is `firmware.hex`

Reviewer: And the file is located in `build/bbc-microbit-classic-gcc-nosd/source` is not in `build` ?

Author: I get another build error

```
Linking CXX executable source/microbit-micropython
tools/adduicr.py build/bbc-microbit-classic-gcc-nosd/source/microbit-micropython.hex $(cat build/veraddr.txt) -o build/firmware.hex
usage: adduicr.py [-h] [-o OUTPUT] firmware address
adduicr.py: error: argument address: invalid <lambda> value: '0x'
make: *** [build/firmware.hex] Error 2
```
but I assume the output is supposed to be at `build/firmware.hex`

Reviewer: Yeah, I'm getting the same issue on my normal dev environment, but not on an old VM. The source code hasn't changed so it might be related to using newer toolchain.
I can confirm the location, can you update the docs and we'll merge the PR?

Author: Is the original text not correct then?

This is still true as a `microbit-micropython.hex` is created in `build/bbc-microbit-classic-gcc-nosd/source`
```
The result is a microbit-micropython hex file
(i.e. ``microbit-micropython.hex``) found in the
build/bbc-microbit-classic-gcc-nosd/source from the root of the repository.
```
As is this as a final 'firmware.hex` file is created in /build
```
The Makefile does some extra preprocessing of the source, which is needed
only if you add new interned strings to qstrdefsport.h. The Makefile also puts
the resulting firmware at build/firmware.hex, and includes some convenience
targets.
```

Reviewer: Yeah, the first block is correct, but mentioning the `microbit-micropython.hex` file in the build instructions can be accidentally interpreted as that being the main hex file to use (the text does **not** say that, but it's common for users to skim over the docs to find "the bit they need" even if that ends up being the wrong bit).
I'd rather remove mentions of the "middle steps" hex files to avoid people accidentally using that instead of the final hex.

>The Makefile does some extra preprocessing of the source, which is needed
only if you add new interned strings to qstrdefsport.h. The Makefile also puts
the resulting firmware at build/firmware.hex, and includes some convenience
targets.

The extra preprocessing now also includes adding the UICR, which is needed for the final hex, so using the makefile is always required now.

Author: Thanks, Carlos. Agreed. I have removed the unnecessary sentence and updated in https://github.com/bbcmicrobit/micropython/pull/711/commits/7d6951b3d59d0b28c472e7a88c2e2302468d0032

Reviewer: I've opened https://github.com/bbcmicrobit/micropython/issues/712 about the build issues

the resulting firmware at build/firmware.hex, and includes some convenience
targets.

Preparing firmware and a Python program
=======================================

Using ``tools/makecombinedhex.py`` you can combine the MicroPython firmware
with a Python script and produce a hex file ready for uploading to the
micro:bit.::

  ./makecombinedhex.py <firmware.hex> <script.py> [-o <combined.hex>]
  # Output goes to stdout if no filename is given.

Using ``tools/hexlify.py`` you can turn a Python script into Intel HEX format
to be concatenated at the end of the MicroPython firmware.hex.  A simple header
is added to the script.::

  ./hexlifyscript.py <script.py>
  
It also accepts data on standard input.

micro:bit V2
============

This applies to MicroPython for the micro:bit V2, the source of which can be 
found here: `microbit-foundation/micropython-microbit-v2 <https://github.com/microbit-foundation/micropython-microbit-v2>`_.

The repository also contains a history of 
`MicroPython firmware builds <https://github.com/microbit-foundation/micropython-microbit-v2/actions>`_.

Dependencies
------------

- `cMake <https://cmake.org/>`_
- `git <https://git-scm.com/>`_ 

> - `python <https://www.python.org/downloads/>`_

Reviewer: also needs arm gcc, and ninja in Windows.

Reviewer: Also, since it is building the cross compiler it needs the "normal" gcc to do that.

Reviewer: Done.

Clone the repository::

>     $ git clone https://github.com/microbit-foundation/micropython-microbit-v2

Reviewer: I guess since we have this now we should also add a "clone step" in the V1 section.

Reviewer: Done

>     $ git clone https://github.com/microbit-foundation/micropython-microbit-v2
> 
> Update the submodules::
> 
>     $ git submodule update --init

Reviewer: I would combine these steps to be able to add the "cd micropython-microbit-v2" as well.

```suggestion
    $ git clone https://github.com/microbit-foundation/micropython-microbit-v2
    $ cd micropython-microbit-v2
    $ git submodule update --init
```
